File: bootstrap/legend-pure-next-bootstrap/legend-pure-next-bootstrap-core/legend-pure-next-bootstrap-execution/src/main/java/org/finos/legend/pure/execution/natives/meta/antlr/AntlrNatives.java
// Copyright 2026 Goldman Sachs
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package org.finos.legend.pure.execution.natives.meta.antlr;

import meta.pure.metamodel.valuespecification.ValueSpecification;
import org.antlr.v4.runtime.ParserRuleContext;
import org.antlr.v4.runtime.tree.ParseTree;
import org.antlr.v4.runtime.tree.TerminalNode;
import org.finos.legend.pure.execution.NativeRepository;
import org.finos.legend.pure.next.parser.GrammarExtension;
import org.finos.legend.pure.execution.NativeRepository.LazyNativeImpl;
import org.finos.legend.pure.execution.NativeRepository.NativeImpl;
import org.finos.legend.pure.execution._E_ValueSpecification;
import org.finos.legend.pure.m3.module.MetadataAccess;

import java.lang.reflect.Field;
import java.lang.reflect.Method;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;
import java.util.function.Function;

/**
 * Bridge between the Pure-side ANTLR primitives (declared in
 * {@code meta::pure::functions::meta::antlr}) and the host-side ANTLR
 * runtime. Registered by {@link NativeRepository}'s {@code registerDefaults}
 * — these are core natives, not opt-in extensions.
 *
 * <p>{@code parseAntlr} dispatches via the per-runtime
 * {@link GrammarExtension} registry (see
 * {@code PureExecution.Builder.withGrammarExtensions}); every other native
 * here is grammar-agnostic — it only knows how to navigate a
 * {@link ParserRuleContext}.</p>
 */
public final class AntlrNatives
{
    private static final ConcurrentHashMap<String, Method> METHOD_CACHE = new ConcurrentHashMap<>();
    /**
     * Cache for {@code ruleNames} field lookup per parser class. Each
     * ANTLR-generated parser exposes a {@code public static final String[] ruleNames};
     * we read it reflectively so {@code grammarRuleName} doesn't need a
     * compile-time reference to any specific parser class.
     */
    private static final ConcurrentHashMap<Class<?>, String[]> RULE_NAMES_CACHE = new ConcurrentHashMap<>();

    private AntlrNatives() {}

    public static void register(Map<String, NativeImpl> natives,
                                Map<String, LazyNativeImpl> lazyNatives,
                                MetadataAccess resolver,
                                NativeRepository repository)
    {
        put(natives, resolver,
            "getText_AntlrContext_1__String_1_",
            args -> ((ParseTree) args.get(0)).getText());

        put(natives, resolver,
            "grammarRuleName_AntlrContext_1__String_1_",
            args -> grammarRuleName((ParserRuleContext) args.get(0)));

        put(natives, resolver,
            "getTopLevelChildren_AntlrContext_1__AntlrContext_MANY_",
            args -> {
                List<ParseTree> children = ((ParserRuleContext) args.get(0)).children;
                List<Object> out = new ArrayList<>(children == null ? 0 : children.size());
                if (children != null) {
                    for (ParseTree c : children) {
                        if (c instanceof ParserRuleContext prc) out.add(prc);
                    }
                }
                return out;
            });

        put(natives, resolver,
            "getChild_AntlrContext_1__String_1__AntlrContext_$0_1$_",
            args -> {
                Object r = invokeNamed(args.get(0), (String) args.get(1));
                return r instanceof ParserRuleContext ? r : null;
            });

        put(natives, resolver,
            "getChildren_AntlrContext_1__String_1__AntlrContext_MANY_",
            args -> {
                Object r = invokeNamed(args.get(0), (String) args.get(1));
                // ANTLR generates `ctx.name()` returning either `XContext` (for
                // single-occurrence sub-rules in the grammar) or `List<XContext>`
                // (for + / *). Both should look like a list from Pure's side so
                // `->map(...)` always works uniformly.
                if (r == null) return List.of();
                if (r instanceof List<?> list) return list;
                return List.of(r);
            });

        put(natives, resolver,
            "getTokenText_AntlrContext_1__String_1__String_$0_1$_",
            args -> {
                Object r = invokeNamed(args.get(0), (String) args.get(1));
                return r instanceof TerminalNode t ? t.getText() : null;
            });

        put(natives, resolver,
            "getTokenTexts_AntlrContext_1__String_1__String_MANY_",
            args -> {
                Object r = invokeNamed(args.get(0), (String) args.get(1));
                if (r instanceof List<?> list) {
                    List<Object> out = new ArrayList<>(list.size());
                    for (Object o : list) {
                        if (o instanceof TerminalNode t) out.add(t.getText());
                    }
                    return out;
                }
                if (r instanceof TerminalNode t) return List.of(t.getText());
                return List.of();
            });

        put(natives, resolver,
            "hasChild_AntlrContext_1__String_1__Boolean_1_",
            args -> invokeNamed(args.get(0), (String) args.get(1)) != null);

        put(natives, resolver,
            "hasToken_AntlrContext_1__String_1__Boolean_1_",
            args -> {
                Object r = invokeNamed(args.get(0), (String) args.get(1));
                return r instanceof TerminalNode;
            });

        put(natives, resolver,
            "getStartLine_AntlrContext_1__Integer_1_",
            args -> (long) ((ParserRuleContext) args.get(0)).getStart().getLine());

        put(natives, resolver,
            "getStartColumn_AntlrContext_1__Integer_1_",
            args -> (long) ((ParserRuleContext) args.get(0)).getStart().getCharPositionInLine() + 1);

        put(natives, resolver,
            "getStopLine_AntlrContext_1__Integer_1_",
            args -> (long) ((ParserRuleContext) args.get(0)).getStop().getLine());

        put(natives, resolver,
            "getStopColumn_AntlrContext_1__Integer_1_",
            args -> {
                ParserRuleContext c = (ParserRuleContext) args.get(0);
                return (long) c.getStop().getCharPositionInLine() + c.getStop().getText().length();
            });

        put(natives, resolver,
            "getChildTextAt_AntlrContext_1__Integer_1__String_$0_1$_",
            args -> {
                ParserRuleContext ctx = (ParserRuleContext) args.get(0);
                int idx = ((Long) args.get(1)).intValue();
                return idx < ctx.children.size() ? ctx.children.get(idx).getText() : null;
            });

        put(natives, resolver,
            "stripTripleQuotesDedented_String_1__String_1_",
            args -> org.finos.legend.pure.next.parser.shared.TripleStringStripper.strip(
                (String) args.get(0)));

        put(natives, resolver,
            "parseAntlr_String_1__String_1__String_1__Integer_1__AntlrContext_1_",
            args -> {
                String source = (String) args.get(0);
                String grammar = (String) args.get(1);
                String sourceId = (String) args.get(2);
                int lineOffset = ((Long) args.get(3)).intValue();
                GrammarExtension ext = repository.lookupGrammar(grammar);
                if (ext == null)
                {
                    throw new RuntimeException("Unknown grammar: " + grammar
                            + " (no GrammarExtension registered; pass it via PureExecution.Builder.withGrammarExtensions)");
                }
                return ext.parse(source, sourceId, lineOffset);
            });

        put(natives, resolver,
            "computeFirstNonNewlineLine_AntlrContext_$0_1$__Boolean_1__Integer_1_",
            args -> {
                Object raw = args.get(0);
                Object ctx = raw instanceof List<?> l ? (l.isEmpty() ? null : l.get(0)) : raw;
                boolean syntheticHeader = (Boolean) args.get(1);
                if (ctx == null) return 0L;
                ParserRuleContext prc = (ParserRuleContext) ctx;
                for (int i = 0; i < prc.getChildCount(); i++)
                {
                    ParseTree child = prc.getChild(i);
                    if (child instanceof TerminalNode tn
                            && tn.getSymbol().getType() == org.antlr.v4.runtime.Token.EOF) continue;
                    String text = child.getText();
                    if (text != null && text.trim().isEmpty()) continue;
                    int line = (child instanceof TerminalNode tn2)
                            ? tn2.getSymbol().getLine()
                            : ((ParserRuleContext) child).getStart().getLine();
                    int offset = line - 1;
                    if (syntheticHeader) offset -= 1;
                    return (long) offset;
                }
                return 0L;
            });
    }

    /**
     * Resolve {@code ctx}'s rule name by walking up to its enclosing parser
     * class (every ANTLR-generated {@code XContext} is an inner class of
     * {@code XParser}) and reading {@code XParser.ruleNames} reflectively.
     * Cached per parser class.
     */
    private static String grammarRuleName(ParserRuleContext ctx)
    {
        Class<?> ctxCls = ctx.getClass();
        Class<?> parserCls = ctxCls.getEnclosingClass();
        if (parserCls == null)
        {
            throw new IllegalStateException(
                    "grammarRuleName: context class " + ctxCls.getName()
                            + " has no enclosing parser class; can't resolve ruleNames");
        }
        String[] ruleNames = RULE_NAMES_CACHE.computeIfAbsent(parserCls, cls ->
        {
            try
            {
                Field f = cls.getField("ruleNames");
                return (String[]) f.get(null);
            }
            catch (NoSuchFieldException | IllegalAccessException e)
            {
                throw new IllegalStateException(
                        "grammarRuleName: " + cls.getName()
                                + " has no public static `ruleNames` field", e);
            }
        });
        return ruleNames[ctx.getRuleIndex()];
    }

    private static void put(Map<String, NativeImpl> natives,
                            MetadataAccess resolver,
                            String signature,
                            Function<List<Object>, Object> body)
    {
        natives.put(signature, (args, eval, genericType, multiplicity) -> {
            List<Object> unwrapped = new ArrayList<>(args.size());
            for (ValueSpecification vs : args) unwrapped.add(_E_ValueSpecification.unwrap(vs));
            Object result = body.apply(unwrapped);
            return _E_ValueSpecification.wrap(result, genericType, multiplicity, resolver);
        });
    }

    private static Object invokeNamed(Object ctx, String methodName)
    {
        try
        {
            Class<?> cls = ctx.getClass();
            Method m = METHOD_CACHE.computeIfAbsent(cls.getName() + "#" + methodName, k ->
            {
                try { return cls.getMethod(methodName); }
                catch (NoSuchMethodException e) { throw new RuntimeException(e); }
            });
            return m.invoke(ctx);
        }
        catch (Exception e)
        {
            throw new RuntimeException("Failed to invoke " + methodName + " on " + ctx.getClass().getSimpleName(), e);
        }
    }
}

File: pure/modules/translation/javascript/js/antlr-natives.js
// ©2026 JP Morgan Chase & Co. All rights reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

// Pure ANTLR-bridge natives — implementations of the 17 functions declared in
// pure/specification/runtime/functions/meta/parser/antlr/*.pure that the
// translated parser-mappings interpreter (parseDocument, buildClass, …) calls
// to walk a ParserRuleContext. Each function mirrors the semantics of the
// Truffle-side AntlrNodes.java helper of the same name so the translated JS
// produces identical PureFile ASTs.
//
// Contexts (ParserRuleContext, TerminalNode) flow opaquely through Pure as
// `meta::pure::functions::meta::antlr::AntlrContext` PDOs; here they are the
// real antlr4 objects, and `ctx[ruleName]()` reflectively reads the generated
// rule-accessor methods (same pattern AntlrNodes uses on the Java-generated
// context classes).

import antlr4 from "antlr4";

import TopLexer from "./generated/TopLexer.js";
import TopParser from "./generated/TopParser.js";
import M3Lexer from "./generated/M3Lexer.js";
import M3Parser from "./generated/M3Parser.js";

const { ParserRuleContext, Token } = antlr4;
const TerminalNode = antlr4.tree.TerminalNode;

// Grammar registry — mirrors the Truffle-side GrammarExtension dispatch.
// Each entry knows how to turn a source string into a top-level
// ParserRuleContext for that grammar. Add new grammars here.
// Run a lexer+parser, collecting syntax errors via a listener that REPLACES the
// default ConsoleErrorListener (which only logs to stderr and lets the parser
// recover silently). If any syntax error fired, throw — so callers see parse
// failures instead of getting a quietly-broken tree. The thrown error carries a
// structured `parseErrors: [{line, column, msg}]` (column 0-based, antlr4
// convention) for editors that want to map errors back to source spans.
function parseWith(LexerClass, ParserClass, rule, source) {
    const lexer = new LexerClass(new antlr4.CharStream(source));
    const parser = new ParserClass(new antlr4.CommonTokenStream(lexer));
    const parseErrors = [];
    const listener = {
        syntaxError(_recognizer, _offendingSymbol, line, column, msg) { parseErrors.push({ line, column, msg }); },
        reportAmbiguity() {},
        reportAttemptingFullContext() {},
        reportContextSensitivity() {},
    };
    lexer.removeErrorListeners();
    lexer.addErrorListener(listener);
    parser.removeErrorListeners();
    parser.addErrorListener(listener);
    const ctx = parser[rule]();
    rememberRuleNames(ctx, parser.ruleNames);
    if (parseErrors.length > 0) {
        const err = new Error(
            "Parse error:\n" + parseErrors.map((e) => `  line ${e.line}:${e.column} ${e.msg}`).join("\n"),
        );
        err.parseErrors = parseErrors;
        throw err;
    }
    return ctx;
}

const GRAMMARS = {
    "TopParser": (source) => parseWith(TopLexer, TopParser, "document", source),
    "M3Parser": (source) => parseWith(M3Lexer, M3Parser, "definition", source),
};

// Each root context is registered with the ruleNames array of the parser that
// produced it. `__grammarRuleName` walks `ctx.parentCtx` up to the root to find
// the matching ruleNames. antlr4 contexts don't carry a `.parser` reference
// directly (the Lexer/Parser instance is short-lived once parse returns), so we
// keep a side table keyed by the root.
const ROOT_RULE_NAMES = new WeakMap();
function rememberRuleNames(root, ruleNames) {
    ROOT_RULE_NAMES.set(root, ruleNames);
}
function ruleNamesOf(ctx) {
    let cur = ctx;
    while (cur) {
        const r = ROOT_RULE_NAMES.get(cur);
        if (r) return r;
        cur = cur.parentCtx;
    }
    throw new Error("ruleNamesOf: context has no registered root (was it produced by __parseAntlr?)");
}

// ============================================================
// parseAntlr — drive ANTLR by grammar name; return the root context.
// ============================================================
export function __parseAntlr(source, grammarName, sourceId, lineOffset) {
    const grammar = GRAMMARS[grammarName];
    if (!grammar) {
        throw new Error("Unknown grammar: " + grammarName + " (known: " + Object.keys(GRAMMARS).join(", ") + ")");
    }
    return grammar(source, sourceId, Number(lineOffset));
}

// ============================================================
// Text + rule-name accessors.
// ============================================================
export function __getText(ctx) {
    return ctx.getText();
}

export function __grammarRuleName(ctx) {
    // Mirrors AntlrNodes.GrammarRuleNameNode: look up the rule name from the
    // enclosing parser's ruleNames table. We stashed it on the root context
    // at __parseAntlr time (antlr4 doesn't carry parser through ctx).
    return ruleNamesOf(ctx)[ctx.ruleIndex];
}

// ============================================================
// Navigation: top-level ParserRuleContext children + reflective accessors.
// ============================================================
export function __getTopLevelChildren(ctx) {
    const children = ctx.children;
    if (!children || children.length === 0) return [];
    const out = [];
    for (const c of children) {
        if (c instanceof ParserRuleContext) out.push(c);
    }
    return out;
}

// Reflective dispatch — same idea as AntlrNodes.invokeNamed(ctx, name): call
// the no-arg accessor method on the context. Generated antlr4 contexts expose
// one method per grammar rule + per token reference.
//
// Some ANTLR targets append `_` to rule names that clash with target-language
// reserved words. The JavaScript target generally does not, but keep the
// fallback so Pure-side parser-mappings code stays grammar-agnostic: try the
// bare name first, then fall back to `name_`.
function invokeNamed(ctx, name) {
    let fn = ctx[name];
    if (typeof fn !== "function") {
        fn = ctx[name + "_"];
    }
    if (typeof fn !== "function") {
        throw new Error("Antlr accessor '" + name + "' missing on " + (ctx.constructor && ctx.constructor.name));
    }
    return fn.call(ctx);
}

export function __getChild(ctx, name) {
    const r = invokeNamed(ctx, name);
    return r instanceof ParserRuleContext ? r : null;
}

export function __getChildren(ctx, name) {
    const r = invokeNamed(ctx, name);
    if (r == null) return [];
    if (Array.isArray(r)) return r;
    return [r];
}

export function __getTokenText(ctx, name) {
    const r = invokeNamed(ctx, name);
    return r instanceof TerminalNode ? r.getText() : null;
}

export function __getTokenTexts(ctx, name) {
    const r = invokeNamed(ctx, name);
    if (r == null) return [];
    if (Array.isArray(r)) {
        return r.map((o) => (o instanceof TerminalNode ? o.getText() : null)).filter((t) => t != null);
    }
    return r instanceof TerminalNode ? [r.getText()] : [];
}

export function __hasChild(ctx, name) {
    return invokeNamed(ctx, name) != null;
}

export function __hasToken(ctx, name) {
    return invokeNamed(ctx, name) instanceof TerminalNode;
}

// ============================================================
// Source position accessors — bigint-typed for Pure Integer compat.
// antlr4's `column` is 0-based; Pure expects 1-based start columns.
// ============================================================
export function __getStartLine(ctx) {
    return BigInt(ctx.start.line);
}

export function __getStartColumn(ctx) {
    return BigInt(ctx.start.column + 1);
}

export function __getStopLine(ctx) {
    return BigInt(ctx.stop.line);
}

export function __getStopColumn(ctx) {
    const stop = ctx.stop;
    return BigInt(stop.column + (stop.text ? stop.text.length : 0));
}

// ============================================================
// getChildTextAt — text of the i-th child (ParserRuleContext or TerminalNode).
// Used by the precedence ladder to read operator tokens.
// ============================================================
export function __getChildTextAt(ctx, idx) {
    const i = Number(idx);
    const children = ctx.children;
    if (!children || i >= children.length) return null;
    return children[i].getText();
}

// ============================================================
// stripTripleQuotesDedented — port of TripleStringStripper.strip (JEP-378).
// Drop the 3-char `'''` delimiters; strip the minimum-leading-whitespace
// column across continuation lines (the last line — where the closing `'''`
// sits — always contributes, even when blank).
// ============================================================
export function __stripTripleQuotesDedented(tripleQuoted) {
    const body = tripleQuoted.substring(3, tripleQuoted.length - 3);
    if (body.indexOf("\n") < 0) return body;

    const leadingNewline = body.charAt(0) === "\n";
    const firstNewline = leadingNewline ? -1 : body.indexOf("\n");
    const firstLine = leadingNewline ? null : body.substring(0, firstNewline);
    const rest = leadingNewline ? body.substring(1) : body.substring(firstNewline + 1);

    const continuationLines = rest.split("\n");

    let minLeading = Number.MAX_SAFE_INTEGER;
    for (let i = 0; i < continuationLines.length; i++) {
        const line = continuationLines[i];
        const isLast = i === continuationLines.length - 1;
        if (!isLast && isBlank(line)) continue;
        const n = countLeadingWhitespace(line);
        if (n < minLeading) minLeading = n;
    }
    if (minLeading === Number.MAX_SAFE_INTEGER) minLeading = 0;

    const out = [];
    if (!leadingNewline) out.push(firstLine);
    for (let i = 0; i < continuationLines.length; i++) {
        if (i > 0 || !leadingNewline) out.push("\n");
        out.push(stripUpTo(continuationLines[i], minLeading));
    }
    return out.join("");
}

function isBlank(s) {
    for (let i = 0; i < s.length; i++) {
        const c = s.charAt(i);
        if (c !== " " && c !== "\t") return false;
    }
    return true;
}

function countLeadingWhitespace(s) {
    let n = 0;
    while (n < s.length) {
        const c = s.charAt(n);
        if (c !== " " && c !== "\t") break;
        n++;
    }
    return n;
}

function stripUpTo(s, n) {
    const strip = Math.min(n, countLeadingWhitespace(s));
    return s.substring(strip);
}

// ============================================================
// computeFirstNonNewlineLine — Section body line-offset computation.
// Walks children, skipping EOF + whitespace-only nodes; returns the
// 0-based line offset of the first content node (adjusted by 1 when the
// host prepended a synthetic ###Pure header).
// ============================================================
export function __computeFirstNonNewlineLine(ctx, syntheticHeader) {
    if (ctx == null) return 0n;
    const childCount = ctx.getChildCount();
    for (let i = 0; i < childCount; i++) {
        const child = ctx.getChild(i);
        if (child instanceof TerminalNode && child.symbol.type === Token.EOF) continue;
        const text = child.getText();
        if (text != null && text.trim().length === 0) continue;
        const line =
            child instanceof TerminalNode
                ? child.symbol.line
                : child.start.line;
        let offset = line - 1;
        if (syntheticHeader) offset -= 1;
        return BigInt(offset);
    }
    return 0n;
}
